### DISCUSSION TITLE
Virtual chin rest plugin

### DISCUSSION THREAD

**Discussion by GEJ1** [jspsych/jsPsych#975] [Show and tell]
Hi everyone,

I was working on wraping up the code from [virtual chin rest]( https://github.com/QishengLi/virtual_chinrest/) [(paper: Qisheng Li et al., 2020)](https://www.nature.com/articles/s41598-019-57204-1) into a plugin, and I think I have a first version that works.

Thank you @jodeleeuw for your super useful youtube video about making plugins: https://youtu.be/XQcsFwAmbiw

If anyone is interested in making improvements:

- code : https://github.com/GEJ1/jsPsych_online_TMT/blob/master/jspsych/plugins/jspsych-virtual-chin.js

- external JS/CSS for using jquery and stimulus : https://drive.google.com/drive/folders/1CP8Cg1xbzILPxejGEXtpLDHQuPxkxFF6?usp=sharing

I used the same external files that the original JavaScript code: https://github.com/QishengLi/virtual_chinrest

- Uploaded in cognition.run: https://9s6ib7pmqv.cognition.run/

ps: Maybe theres is some comments in spanish, sorry hehe.

Best!

Gustavo.

**Comment by Stajduhar-A** [comment]
Very interesting and would love to check it out but I'm greeted with this when I try to view it on cognition.run... 
<img width="1003" alt="Screen Shot 2020-08-13 at 4 59 26 PM" src="https://user-images.githubusercontent.com/52038520/90186611-6d759500-dd86-11ea-95cc-db33353c388a.png">

**Reply by GEJ1** [reply]
Ohh, maybe is the browser? It works for me on Google Chrome

**Reply by jodeleeuw** [reply]
Worked for me on Chrome as well.

**Comment by jodeleeuw** [comment]
This is great! If you are willing to write documentation and create an example for using the plugin we can incorporate it into a future release. You could open a Pull Request to merge this in. We can also work on the documentation and example ourselves, but that'll be slower.

We may also want to consider merging some of this work with the `jspsych-resize` plugin. That plugin does the same work as the first part of the chinrest script, so it would make sense to get rid of the resize plugin and combine these together, perhaps making the blind spot detection part of the task optional.

**Reply by GEJ1** [reply]
Thank you @jodeleeuw ! I'll be glad to contribute making a Pull Request.

**Reply by GEJ1** [reply]
I just opened a Pull Request: https://github.com/jspsych/jsPsych/pull/1024

**Comment by yuvalharr** [comment]
Thats great Gustavo, thank you very much! I just found the original paper and code and hoped so much that someone would be able to make it into a plugin :)

Does your plugin allow to take the scale from the credit card and use it as a measure to resize the rest of the experiment into a certain size, like the Resize plugin?

Anyway, this plugin looks amazing, so thanks again!

Yuval

**Reply by GEJ1** [reply]
Thank you! The plugin is not doing the resize at this moment. My interim solution is to use the jsPsych-resize plugin before the virtual chin in order to resize the screen. 

As @jodeleeuw mention, it would be a good idea to merge the virtual chin and the resize plugins.

**Reply by yuvalharr** [reply]
Thanks for the answer!

So you mean to do twice a credit card measuring- each in every trial? Or were you able to pass the data of the measure from the resize plugin into the virtual chin plugin without the need to make participants measure their card again?

**Reply by GEJ1** [reply]
Yes, I mean  to do twice a credit card measuring, using first  the jsPsych-resize plugin. Doing that at beginning of the experiment it's enough.

**Comment by QishengLi** [comment]
@GEJ1 Thanks for your interest in our work and making effort to convert it to a jsPsych plugin! Please let me know how I can help - I am happy to review/contribute to the pull request.

Best,
Qisheng

**Reply by GEJ1** [reply]
Thank you @QishengLi ! I think it is a really useful tool what you made. I'll be posting here the news about the plugin in order to colaborate.

Best,

Gustavo.

**Comment by jodeleeuw** [comment]
Thanks for opening the pull request! 

I do think it would be worth merging some of the resize plugin features into this, like being able to select the object size for scaling and having the rest of the experiment scaled to a particular size. We should probably also make the text configurable so it can be used in other languages. One way to do this is to make the instructions for using the plugin a parameter.

Other really useful things for the pull request would be:

* updates to the documentation
* an example in the examples folder
* tests in the tests folder as appropriate
* updates to the contributors.md file in the main folder

We should make sure that the documentation page links to @QishengLi 's paper and provides the appropriate citation info. The documentation for the RDK plugin is a nice example of this.

**Comment by yuvalharr** [comment]
Sorry for the many questions, but I still have some issues.

After the virtual chinrest plugin is over, it doesn't continue to the next trial (which is an html-keyboard-response trial). It instead shows an error message right after pressing the space bar:

_jspsych-html-keyboard-response.js:109 Uncaught TypeError: Cannot read property 'className' of null
    at Object.after_response [as callback_function] (jspsych-html-keyboard-response.js:109)
    at Object.listener_function [as fn] (jspsych.js:2064)
    at HTMLBodyElement.root_keydown_listener (jspsych.js:1983)_

Couldn't manage to understand what might cause this, as it is working fine without the chinrest plugin. My best guess is something to do with the additional css files that I might didn't use them right.

Here is part of my html page <head> where I put all of the files you provided with the plugin:

```
    <script src="jspsych-6.1.0/plugins/jspsych-virtual-chin.js"></script>
    <script src= "jspsych-6.1.0/jquery.min.js"></script>
    <script src= "jspsych-6.1.0/jquery-ui.min.js"></script>
    <script src= "jspsych-6.1.0/svg.min.js"></script>
    <link href="jspsych-6.1.0/css/jspsych.css" rel="stylesheet" type="text/css"></link>
    <link href="jspsych-6.1.0/css/bootstrap.min.css" rel="stylesheet" type="text/css"></link> 
    <link href="jspsych-6.1.0/css/jquery-ui.css" rel="stylesheet" type="text/css"></link>
    <link href="jspsych-6.1.0/css/styles.css" rel="stylesheet" type="text/css"></link>
```
Does it look OK? Any ideas to what might be the problem?

Thanks very much,
Yuval

**Reply by GEJ1** [reply]
Hi @yuvalharr , sorry for the delay. I made a commit today and I think now it's fixed, please tell me if not: https://github.com/GEJ1/jsPsych/blob/master/plugins/jspsych-virtual-chin-rest.js

The functionality for going to next trial was working weird, in my experiment was working but I think know works with every plugin at next.

Best!

Gustavo.

**Reply by yuvalharr** [reply]
Hi, thank you for your time and effort!

Actually I've tried your fix but it still has some problems... In my experiment, after the chin-rest trial it goes to an html-keyboard-response trial, which should include several html texts one after the other, advancing to the next message after every space bar press. The problem is that it gets stuck at the first stimulus, and pressing the space bar doesn't work in order to advance to the next html-keyboard-response stimulus.

At this point I've actually found a fix that was OK for me. I just put a different trail that turns out to work (animation), for only milliseconds, right after your chin-rest trial. Turns out that when putting it between the chin-rest and the html-keyboard-response it makes the html-keyboard-response trial work well. So don't bother for me, it works good for me now :)

Again, thank you very much for your help and for this great plugin!

Yuval

**Reply by GEJ1** [reply]
Excellent! Tell thanks to @QishengLi , she made the plugin, I'm just helping to put it in the jsPsych structure. Hope we can make it available as a well-functioning plugin soon.

If anyone wants to contribute, I already opened a pull request with the code and the documentation: https://github.com/jspsych/jsPsych/pull/1024 

Best! :) 

Gustavo.

**Comment by pjkohler** [comment]
Hi Gustavo,

I had somehow missed your implementation so spent some time today making my own. It looks like we've taken pretty similar strategies when implementing @QishengLi's code. I was mainly trying to touch as little of that code as possible, and then simply build a JsPsych plug-in around it. My main interest is having a self-contained plugin that can give me the viewing distance and px2mm, which I can then use to compute approximate px2deg. I am not so interested in having resizing built into the function, as I find it more useful to define the size of individual display elements in physical units. Would you allow me to contribute to your plugin, and if so how do I go about it?

Best,

Peter

**Reply by GEJ1** [reply]
Hi @pjkohler ,

Of course you can contribute with the implementation of the @QishengLi 's code, you can find the PR [here](https://github.com/jspsych/jsPsych/pull/1024) 
My main interest is to help the community with the most general purpose plugin as possible  (maybe merging with the resize plugin) in order to incorporate it into a future release of jsPSych.

Best!

Gustavo.

**Reply by pjkohler** [reply]
Okay, great. I've made some changes to the code. How do I share them with you (sorry, not Github superuser)?

Best,

Peter

**Reply by GEJ1** [reply]
I think you can made a change in my pull request (not github superuser needed). Actually it's the first time that I made contribute to an open source project. Maybe @jodeleeuw  or @becky-gilbert can tell us what would be the best way to do it ?

**Reply by jodeleeuw** [reply]
It looks like the PR was merged in October, but the code isn't in the main branch of the repository currently. I'm not sure what happened, to be honest. 

I think what would make sense is to collaborate on the plugin in a fork, either @GEJ1's fork or a new fork that is up-to-date. You can submit PRs to the fork, and then we can merge the completed plugin into the main branch. 

If you are able to get this ready within the next two weeks we should be able to include it as part of the 6.3 release. Otherwise we can just add it to a future release.

My 2 cents about the plugin design: I do think it make sense to merge the resize plugin features with this plugin, but I would totally support the various features being optional so that it fits the use case Peter is looking for. What do you think about that approach?

**Reply by pjkohler** [reply]
@jodeleeuw: I like that general approach, but we might need some help incorporating the resize plugin. I was confused about the fact that the code wasn't in the main branch, happy that it was not just me.

@GEJ1: Should I push my changes to your existing fork, or do you want to make a new one?

**Reply by jodeleeuw** [reply]
Happy to help as needed!

**Reply by GEJ1** [reply]
Thank you @jodeleeuw ! 

@pjkohler: I have no problem if you want to push my existing fork, but if it's easier for you to create a new one, feel free to do so. I agree that incorporating the resize plugin may not be so straightforward.

**Reply by pjkohler** [reply]
I cloned this fork https://github.com/GEJ1/jsPsych, added my changes, but I think you need to give me permission to push.

**Reply by pjkohler** [reply]
Thanks, pushed my changes. Let me know if I have done anything terrible.

**Reply by pjkohler** [reply]
@jodeleeuw: we just made a new pull request from @GEJ1's fork that incorporates some new functionality including optional resizing.
